worker/src/main/java/gov/cms/ab2d/worker/processor/JobProcessorImpl.java

```diff
@@ -308,23 +555,99 @@ private void doDelete(Path path) {
+     * @return - the mapping
+     */
+    private ContractMapping getContractMappingFromFuture(Future<ContractMapping> future, String contractId) throws ExecutionException, InterruptedException {
+        try {
+            return future.get();
+        } catch (InterruptedException | ExecutionException e) {
+            log.warn("InterruptedException while calling Future.get() - Getting Mapping for " + contractId);
+            throw e;
+        } catch (CancellationException e) {
+            // This could happen in the rare event that a job was cancelled mid-process.
+            // due to which the futures in the queue (that were not yet in progress) were cancelled.
+            // Nothing to be done here
+            log.warn("CancellationException while calling Future.get() - Getting Mapping for " + contractId);
+        }
+        return null;
+    }
+
+    private List<PatientDTO> addDateRangeToExistingOrNewPatient(ContractBeneficiaries beneficiaries,
+                                                    FilterOutByDate.DateRange monthDateRange, Set<String> bfdPatientIds) {
+        if (beneficiaries == null) {
+            return new ArrayList<>();
+        }
+        List<PatientDTO> newPatients = new ArrayList<>();
+        Map<String, PatientDTO> patientDTOMap = beneficiaries.getPatients();
+        for (String patientId : bfdPatientIds) {
+            PatientDTO patientDTO = patientDTOMap.get(patientId);
+
+            if (patientDTO != null) {
+                // patient id was already active on this contract in previous month(s)
+                // So just add this month to the patient's dateRangesUnderContract
+                if (monthDateRange != null) {
+                    patientDTO.getDateRangesUnderContract().add(monthDateRange);
+                }
+            } else {
+                // new patient id.
+                // Create a new PatientDTO for this patient
+                // And then add this month to the patient's dateRangesUnderContract
+
+                patientDTO = PatientDTO.builder()
+                        .patientId(patientId)
+                        .dateRangesUnderContract(new ArrayList<>())
+                        .build();
+                newPatients.add(patientDTO);
+                if (monthDateRange != null) {
+                    patientDTO.getDateRangesUnderContract().add(monthDateRange);
+                }
+                beneficiaries.getPatients().put(patientId, patientDTO);
+            }
+        }
+        return newPatients;
+    }
+
+    private Date getStartDate(String contract) {
+        SimpleDateFormat sdf = new SimpleDateFormat("MM/dd/yyyy");
+
+        String dateToUse = startDate;
+        if (isContractSpecial(contract)) {
+            dateToUse = startDateSpecialContracts;
+        }
+        Date date;
+        try {
+            date = sdf.parse(dateToUse);
+        } catch (ParseException e) {
+            LocalDateTime d = LocalDateTime.of(2020, 1, 1, 0, 0, 0, 0);
```

Comment:
Should this be current year instead of hardcoding it to 2020?

Reply:
No, it's the earliest date we are allowing data to be returned. Next year, it will also be 1/1/2020.